### ISSUE TITLE
Annotating severely occluded objects

### ISSUE THREAD

**Issue by higo522** [ultralytics/ultralytics#21447]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I am currently annotating aerial imagery (extracted from video) of animals under canopy cover, and encounter many partially occluded and truncated animals. I am currently using the following rules:
1. Objects where you can see edges but the middle is occluded the bounding box should enclose the entire object, including the partially occluded section
2. When the rest of the object is completely occluded or cut off by the edge of the frame, the bbox should only cover the visible section (without guessing the full extent that is not visible)
3. If you, as the human expert, cannot confidently identify its class (**without** using prior knowledge from previous frames to make that call), do not annotate it. 

Regarding rule 3, I am facing a dilemma. My goal is to create a high-precision species confirmation model, and I am focusing on minimizing FPs (misclassifying the animal species) and maximizing classification accuracy. FNs (missing the animals) is a lesser concern. Thus, I want to enforce a stricter criterion for prediction, so the model can only give clean, high confidence/accuracy detections of the target classes. 
With this in mind, I am stuck between the following:
- Treat significantly occluded or truncated objects where I cannot confidently identify its class (without prior knowledge) as background, by not annotating them. 
- Delete these ambiguous cases from the dataset altogether to avoid training on confusing features 

However, I also want the model to be able to robustly detect occluded animals under canopy cover, while still ignoring ambiguous cases to suppress FPs and ensure it only gives clean, highly confident predictions (since I have many consecutive frames). I want to avoid giving YOLO data that is only going to confuse the training process since the partially occluded objects have no visual features that suggest a class. 

I have excluded frames with truncated objects at the frame edge from the training dataset (to ensure high quality samples for training). However, is it better to include them and treat them as background? I feel like this would be "lying" to the model since I am teaching it to consider moose features as background. 
Moreover, for animals that are severely occluded by vegetation and I cannot confidently identify its species, should I treat it as background, or exclude this frame?
Lastly, for frames that contains both a fully visible animal and a severely occluded/truncated animal, should I include this frame and only annotate the fully visible individual? (treat the other as background). Or delete this from the dataset?

I would greatly appreciate any thoughts on my annotation strategy and which method I should use for ambiguous cases.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @higo522, thank you for reaching out and for your detailed description of your annotation strategy and dilemmas! 🦌🌳 This is an automated response to help streamline your experience. An Ultralytics engineer will review your question and provide further assistance soon.

We recommend visiting the [Docs](https://docs.ultralytics.com/) for comprehensive guides and best practices. For annotation and custom training questions like yours, please provide as much detail as possible, including sample annotated images and, if available, training logs. You can also review our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) to enhance your project.

If this is a 🐛 Bug Report, please include a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us assist you faster.

Join the Ultralytics community for more insights and real-time discussion:
- [Discord](https://discord.com/invite/ultralytics) 🎧
- [Discourse](https://community.ultralytics.com/)
- [Subreddit](https://www.reddit.com/r/Ultralytics/)

## Upgrade

To ensure you have access to the latest features and fixes, upgrade to the newest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

You can run YOLO in any of these up-to-date verified environments (with dependencies like [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks**: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud**: [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon**: [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you again for your thoughtful question and for using Ultralytics! An engineer will be with you shortly to provide more targeted guidance on your annotation approach.

**Comment by glenn-jocher** [comment]
For your high-precision species confirmation goal, I'd recommend **excluding ambiguous cases entirely** rather than treating them as background. Your instinct about "lying to the model" is correct - labeling partial animal features as background can confuse training and hurt performance.

Here's my suggested approach: Keep frames with mixed visibility but only annotate the clearly identifiable animals, leaving severely occluded ones unannotated. This teaches the model to be confident about clear cases while not providing conflicting signals about animal features. For frames with only ambiguous/truncated animals, exclude them completely to maintain dataset quality.

This aligns with your precision-focused objective and follows the principle that [quality data is the foundation of a successful project](https://docs.ultralytics.com/guides/data-collection-and-annotation/). Consistent application of these rules across your dataset will be more beneficial than trying to squeeze training value from ambiguous cases that could introduce noise.

**Comment by higo522** [comment]
@glenn-jocher I greatly appreciate your response. 

Let me confirm one more thing: if one animal is fully within the frame and another animal is significantly **truncated** at the frame edge within the same photo, I would keep this frame (without deleting it) and only annotate the animal within the frame and leave the cut-off animal unannotated (as background), correct? 

Also, I am thinking of keeping the ambiguous frames (animals severely occluded by canopy cover) in a separate "hard example" dataset, where I can test the model (which was trained only on selected, high-quality, fully identifiable animals) to see its performance for these difficult cases (ex. "--% of heavily occluded animals were successfully detected/classified by the model").

Can I ask for your thoughts on this testing process? 
Thank you in advance.

**Comment by higo522** [comment]
> For your high-precision species confirmation goal, I'd recommend **excluding ambiguous cases entirely** rather than treating them as background. Your instinct about "lying to the model" is correct - labeling partial animal features as background can confuse training and hurt performance.
> 
> Here's my suggested approach: Keep frames with mixed visibility but only annotate the clearly identifiable animals, leaving severely occluded ones unannotated. This teaches the model to be confident about clear cases while not providing conflicting signals about animal features. For frames with only ambiguous/truncated animals, exclude them completely to maintain dataset quality.
> 
> This aligns with your precision-focused objective and follows the principle that [quality data is the foundation of a successful project](https://docs.ultralytics.com/guides/data-collection-and-annotation/). Consistent application of these rules across your dataset will be more beneficial than trying to squeeze training value from ambiguous cases that could introduce noise.

@glenn-jocher Also, one more question if you don't mind. I am training a separate model with the opposite goal: a high-recall sensitivity model that can pick up heavily occluded/truncated animals (with minimal FNs). In this case, I would annotate ambiguous animal features (truncated/occluded) to teach the model to be sensitive; however there are frames with animals that are indistinguishable from background noise and would be unrealistic to train the model on. Since I still need a certain level of accuracy to avoid producing an unmanageable amount of FPs, would you also recommend **deleting** this frame from training instead**treating it as background** to avoid "lying" to the model and giving conflicting signals about animal features? 

Thank you very much.

**Comment by glenn-jocher** [comment]
Yes, even for your high-recall model I'd recommend **deleting** those indistinguishable cases rather than treating them as background. The same "lying to the model" principle applies - if there are truly no distinguishable animal features visible, labeling those pixels as background creates conflicting training signals.

For the sensitivity model, you'd want to include heavily occluded cases where there's still *some* visual evidence of an animal (partial body parts, characteristic shapes, etc.), but draw the line at cases that are genuinely indistinguishable from noise. This maintains the model's ability to learn what minimal animal features look like while avoiding training on pure noise patterns that would hurt both precision and meaningful recall.

**Comment by higo522** [comment]
@glenn-jocher Thank you. For frames with **both** distinguishable and indistinguishable animals, would you say I should keep this frame and only annotate the former and keep the latter unannotated (and treat as background)? Thank you!

**Comment by glenn-jocher** [comment]
Yes, exactly - keep the frame and only annotate the distinguishable animals, leaving the indistinguishable ones unannotated. This approach maximizes your training data while maintaining quality, since you're not actively mislabeling animal features as background but simply not providing labels where identification isn't confident.

**Comment by higo522** [comment]
@glenn-jocher Thanks again for your advice. My final question is on the best practice for annotating **truncated objects** at the frame edge for a high-precision model. Given that YOLO's own augmentations (like mosaic) create many random truncations during training, what is the most effective strategy for handling truncated objects in the source data?

A) The "Identifiability" Rule: Treat truncation exactly like occlusion. Annotate a truncated part only if it's clearly identifiable (e.g., a head), and **delete frames** with unidentifiable parts (e.g., hind legs) unannotated. In a mixed-visibility frame, **keep this frame** and only annotate the animal within the frame (or truncated but still identifiable) and treat the severely truncated (and ambiguous; ex. hind legs) as background. 

B) The "Annotate All" Rule: Annotate all visible truncated parts, regardless of identifiability, to provide the model with the maximum number of real-world partial examples.

Your insight would be greatly appreciated. Regarding A), especially for the mixed cased scenario, treating the truncated animal part (ex. horses legs) as background still feels like I am "lying" to the model by considering a perfectly visible horse leg (although not enough to identify a horse from a deer, cow, etc. on its own) as the background... Wouldn't this also be "mislabeling animal features as background" and give conflicting signals? This is an example frame of a mixed case scenario. I have 3 options:
1. Delete this frame 
2. Annotate both
3. Only annotate the fully visible right and leave the left (truncated) unannotated, treating it as background

<img width="497" height="201" alt="Image" src="https://github.com/user-attachments/assets/0f964520-ffe8-46c6-873c-2a86908e6c15" />

**Comment by glenn-jocher** [comment]
You're absolutely right to be concerned about "lying to the model" - treating clearly visible animal features like legs as background would indeed create conflicting training signals.

For your high-precision model, I'd recommend **option 2: annotate both** in cases like your example where the truncated portion still shows clear animal features (legs, body parts, etc.). The key distinction is whether there are *identifiable animal features* present, not whether you can determine the exact species from that portion alone.

Since YOLO's augmentations will create similar truncations during training anyway, providing these real truncation examples with proper labels actually helps the model learn to handle edge cases correctly. Reserve frame deletion for cases where the truncated portion truly shows no distinguishable animal characteristics or is just pure noise.

The "Identifiability" rule works better when applied as "Are there visible animal features?" rather than "Can I identify the species?" - this maintains consistency with your precision goals while avoiding the conflicting signals you're worried about.

**Comment by higo522** [comment]
Thank you so much for the clarirfication, this helps immensely. This drastically changes my annotation strategy; I have a few final follow-up questions to ensure I apply it correctly:

**1. Extreme Truncation & Precision:**
Consider this mixed frame with one full horse and another severely truncated, like only the left edge of its body is visible. Is the correct action to annotate both, as long as I can tell that its not part of the background and is SOME animal part (regardless of its species)? My high-precision, multi-species model requires classification of very similar signatures. Should I trust the model to learn from such a generic partial feature without harming its classification precision?

<img width="1237" height="355" alt="Image" src="https://github.com/user-attachments/assets/4cfb5bc0-b401-406c-9fdb-5b5a880a3756" />

**2. Consistency with Occlusion:**
Should this rule—annotating any "distinguishable animal feature" (not identifiable exact species)—apply equally to occlusion? For instance is the following correct?:
A) If the entire horse body is obscured by thick canopy and is diffficult to distingusih from background, I treat it as background (leave unannotated)
B) If even a small part of the body is sticking out and clearly visible (ex. horses legs) but is not enough to identify the species alone, I annotate it with its species label

For example in the following case, do I leave the right horse (covered by canopy, red arrow) unannotated, or should I label it?

<img width="1071" height="475" alt="Image" src="https://github.com/user-attachments/assets/bb399b63-0aaa-4d4c-833c-d6139da2f55e" />

**3. A Broader Question:**
This leads me to a final thought. If the model naturally assigns higher confidence scores to complete objects, I can achieve my high-precision goal by adjusting the confidence threshold during inference.
Given this, should I revise the earlier advice and now include standalone frames of only ambiguous/truncated animals in the training set (instead of deleting them to maintain dataset quality)? Would this provide more data for better results, trusting the confidence threshold to handle precision later? In short, should I treat standalone and mixed frames equally, labeling ALL distinguishable animal features, whether its the only thing in the frame or mixed with a fully visible animal?

Your insight on these final points would be invaluable.

**Comment by glenn-jocher** [comment]
**1. Extreme Truncation:** For cases like your severely truncated example, if only a thin edge of the body is visible without any distinctive animal characteristics, I'd lean toward leaving it unannotated. The key test is whether there are enough meaningful features for the model to learn from - a generic body edge provides minimal learning value and could introduce noise.

**2. Consistency with Occlusion:** Yes, apply the same "distinguishable animal features" rule. In your canopy example, if the horse under the red arrow is so obscured that you can't distinguish clear animal features from background vegetation, leave it unannotated. But if you can see distinct legs, body contours, or other clear animal parts, annotate it.

**3. Broader Question:** While confidence thresholds can help during inference, I'd still recommend maintaining the current strategy of excluding standalone ambiguous frames. Quality training data remains more important than quantity - even with confidence filtering, poor training examples can degrade the model's feature learning. Keep focusing on frames where you can identify clear animal features, whether standalone or mixed.

The principle remains: annotate when there are meaningful, distinguishable animal characteristics present, but avoid cases that are pure noise or provide minimal learning signal.

**Comment by higo522** [comment]
Thank you for the final clarification. Your advice has been invaluable. 
Regarding the "distinguishable animal feature" rule, my last question is on consistency.

Previously, I was using a stricter threshold for standalone frames (deleting them if the exact species wasn't identifiable) to maximize data quality. However, for mixed frames, I used a more lenient threshold (annotating any visible leg, etc.) to avoid the "conflicting signal" of treating real body features as background, as you have advised. 

Should I abandon this two-tiered system and apply the same rule everywhere?

For example, should I now keep and annotate a standalone frame even if it only contains a single, non-specific hind leg, just as I would in a mixed frame?

Thank you very much.

**Comment by glenn-jocher** [comment]
Yes, you should apply the same "distinguishable animal features" rule consistently across both standalone and mixed frames for better training consistency.

However, refine the threshold to focus on **meaningful** distinguishable features rather than just any animal part. A single generic hind leg in isolation (whether standalone or mixed) likely doesn't provide enough distinctive characteristics for effective learning. 

The key distinction isn't standalone vs mixed frames, but rather whether the visible features contain enough meaningful animal characteristics to contribute positively to training. Apply this same standard everywhere - if you wouldn't consider a feature meaningful enough to learn from in a mixed frame, don't include it in standalone frames either.

This maintains consistency while ensuring you're only training on features that actually help the model learn robust animal detection patterns.

**Comment by higo522** [comment]
Thank you as always for your help. I have another question:

Black Borders: Many of my frames have varying-sized black borders from camera zoom. Will these empty pixels negatively impact training, and if so should I exclude/delete this frame from the training set? Or can I include this trusting that the model is robust enough to treat them as background padding?

<img width="875" height="329" alt="Image" src="https://github.com/user-attachments/assets/978b0d1d-933c-4ab4-86f2-cca9b003c0dc" />

Slight Blur/Glitches: My operational goal for the model is to be intended for use on clear imagery (its detections on severely blurred/glitched frames are useless). However, some frames have slight motion blur or glitches (white lines on screen) where the animal's features still remain clearly distinguishable. Should I still include these valuable examples to improve robustness, or should I strictly delete any frame that isn't perfectly clear?

**Comment by glenn-jocher** [comment]
Black borders are fine to include—models effectively treat them like padding; only bother cropping if they’re large and you want to save compute. For blur/glitches, keep frames where animal features are still clearly distinguishable and drop heavily blurred/artifacted ones; mild cases are well handled by training augmentations—see the Data Augmentation using Ultralytics YOLO docs for details.

**Comment by higo522** [comment]
Thank you so much for your support so far. It has been pivotal to my project. 
I have another specific question on annotating bunched animals. My primary goal is species ID, not counting.

For extreme cases where deer merge into a single thermal blob, making it impossible to distinguish individuals (example image below showing 3 clumped up deer), what is the correct approach?

<img width="281" height="267" alt="Image" src="https://github.com/user-attachments/assets/97ea89d7-f5f9-4e52-959c-c187d4db7a3f" />

A) Annotate the entire blob with a single deer label.
B) Only annotate individuals I can discern (without using prior knowledge from previous frames): in this case, 2 deer
C) Rigorously annotate every single individual completely even when it is impossible to tell from that frame alone: in this case, 3 deer

Thank you for the clarification!

**Comment by glenn-jocher** [comment]
Go with B: annotate only the individuals you can clearly separate in that frame, and leave the fused blob unannotated—labeling the whole blob as a single deer (A) teaches the wrong prior, and inferring hidden individuals (C) injects label noise; if blobs are common and you need to flag them, consider a separate “group” class or handle those cases with segmentation, and keep your rules consistent as noted in the [Data Collection and Annotation Strategies](https://docs.ultralytics.com/guides/data-collection-and-annotation/).
